File: 0935-knight-dialer/0935-knight-dialer.java
class Solution {
    int mod = 1000000000 + 7;
    long findPath(int i, int j, int n, long[][][] dp) {
        if(i < 0 || j < 0 || i >= 4 || j >= 3 || (i == 3 && j != 1)) return 0;
        if(n == 1) return 1;
        if(dp[i][j][n] > 0) return dp[i][j][n];
        
        dp[i][j][n] = (findPath(i - 1, j - 2, n - 1, dp) % mod + 
                           findPath(i - 1, j + 2, n - 1, dp) % mod + 
                           findPath(i + 1, j - 2, n - 1, dp) % mod + 
                           findPath(i + 1, j + 2, n - 1, dp) % mod + 
                           findPath(i - 2, j - 1, n - 1, dp) % mod + 
                           findPath(i - 2, j + 1, n - 1, dp) % mod + 
                           findPath(i + 2, j - 1, n - 1, dp) % mod + 
                           findPath(i + 2, j + 1, n - 1, dp) % mod) % mod;
        
        return dp[i][j][n];
    }
    
    public int knightDialer(int n) {
        long ans = 0;
        long[][][] dp = new long[4][3][n+1];
        for(int i = 0; i<4; i++) {
            for(int j = 0; j<3; j++) {
                ans = (ans + findPath(i, j, n, dp)) % mod;
            }
        }
        
        return (int) ans;
    }
}

File: 0935-knight-dialer/0935-knight-dialer.py
class Solution(object):
    def __init__(self):
        self.mod = 10**9 + 7

    def find_path(self, i, j, n, dp):
        if i < 0 or j < 0 or i >= 4 or j >= 3 or (i == 3 and j != 1):
            return 0
        if n == 1:
            return 1
        if dp[i][j][n] != -1:
            return dp[i][j][n]

        dp[i][j][n] = (
            self.find_path(i - 1, j - 2, n - 1, dp) % self.mod +
            self.find_path(i - 1, j + 2, n - 1, dp) % self.mod +
            self.find_path(i + 1, j - 2, n - 1, dp) % self.mod +
            self.find_path(i + 1, j + 2, n - 1, dp) % self.mod +
            self.find_path(i - 2, j - 1, n - 1, dp) % self.mod +
            self.find_path(i - 2, j + 1, n - 1, dp) % self.mod +
            self.find_path(i + 2, j - 1, n - 1, dp) % self.mod +
            self.find_path(i + 2, j + 1, n - 1, dp) % self.mod
        ) % self.mod

        return dp[i][j][n]
    
    def knightDialer(self, n):
        ans = 0
        dp = [[[ -1 for _ in range(n + 1)] for _ in range(3)] for _ in range(4)]

        for i in range(4):
            for j in range(3):
                ans = (ans + self.find_path(i, j, n, dp)) % self.mod

        return int(ans)
        